relay-auth/src/lib.rs

```diff
@@ -687,9 +698,113 @@ impl RegisterResponse {
+    /// Bytes of an envelope body that need are used to produce a signature.
+    Body(Bytes),
+    /// No data is needed for this signature because we only want to see if
+    /// the receiving relay can verify the signature correctly.
+    RelayEnvelopeSign,
+}
+
+impl TrySign {
+    /// Creates a signature based on the variant.
+    ///
+    /// Some variants might not produce a signature if credentials are missing, in which case
+    /// `None` is returned.
+    ///
+    /// If credentials are mandatory, it will return `RelaySignatureError::MissingCredentials`.
+    pub fn create_signature(
+        self,
+        secret_key: Option<&SecretKey>,
+    ) -> Result<Option<String>, RelaySignatureError> {
+        match self {
+            TrySign::Body(data) => {
+                let credentials = secret_key.ok_or(RelaySignatureError::MissingCredentials)?;
+                Ok(Some(credentials.sign(data.as_ref())))
+            }
+            TrySign::RelayEnvelopeSign => {
+                let Some(credentials) = secret_key else {
+                    return Ok(None);
+                };
+                Ok(Some(credentials.sign(&[])))
+            }
+        }
+    }
+}
+
+/// Contains information necessary to verify the correctness of the signature.
+#[derive(Debug, Clone, PartialEq)]
+pub struct RelaySignatureData {
+    /// The signature string.
+    pub signature: String,
+    /// The data which the signature will be checked against.
+    pub data: Bytes,
+}
+
+impl RelaySignatureData {
+    /// Creates a new [`RelaySignatureData`] container.
+    pub fn new(signature: String, signature_data: Bytes) -> Self {
+        Self {
+            signature,
+            data: signature_data,
+        }
+    }
+
+    /// Verifies the signature against a single public key and checks if the timestamp
+    /// is within the specified `max_age`.
+    pub fn verify_timestamp(&self, public_key: &PublicKey, max_age: Option<Duration>) -> bool {
+        public_key.verify_timestamp(&self.data, &self.signature, max_age)
+    }
+
+    /// Verifies the signature against a single public key.
+    pub fn verify(&self, public_key: &PublicKey) -> bool {
+        self.verify_timestamp(public_key, None)
+    }
+
+    /// Verifies the signature against any of the public keys and returns `true` if
+    /// one of them can verify it successfully. Also makes sure that the timestamp
+    /// is within the specified `max_age`.
+    pub fn verify_any_timestamp(
+        &self,
+        public_keys: &[PublicKey],
+        max_age: Option<Duration>,
+    ) -> bool {
+        public_keys
+            .iter()
+            .any(|public_key| self.verify_timestamp(public_key, max_age))
+    }
+
+    /// Verifies the signature against any of the public keys and returns `true` if
+    /// one of them can verify it successfully.
+    pub fn verify_any(&self, public_keys: &[PublicKey]) -> bool {
+        public_keys.iter().any(|public_key| self.verify(public_key))
+    }
+}
```

Comment:
Nit: It's kind of messy to have one set of functions taking no timestamp and another set taking an optional timestamp. Either make separate functions with/without timestamps or just one set with optional timestamps.